Table: add TableOrganizer and ShowInvisibleColumnsForm for Scout JS

bschwarzent opened the pull request.

mvilliger commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/basic/table/organizer/ShowInvisibleColumnsForm.java:
```diff
@@ -100,7 +89,7 @@ protected int getConfiguredGridColumnCount() {
     protected void execInitField() {
       super.execInitField();
-      setStatusVisible(false);
+      //setStatusVisible(false);
```
Is this no longer required? Then remove the line? Otherwise remove comment?

bschwarzent replied:
Yes, this is no longer required (table status is already invisible on the table field). Method was removed.

mvilliger commented on eclipse-scout-core/src/table/TableModel.ts:
```diff
@@ -268,4 +268,5 @@ export interface TableModel extends WidgetModel {
    */
   textFilterEnabled?: boolean;
   defaultMenuTypes?: string[];
+  tableOrganizer?: ObjectOrModel<TableOrganizer>;
```
if a tableOrganizer model is passed as part of this TableModel, is it actually used when creating the instance?
Either it should have an effect or here only a TableOrganizier can be passed and no model?

bschwarzent replied:
Type was changed to `TableOrganizer` only, because the organizer does not even have a model.

mvilliger commented on eclipse-scout-core/src/table/TableAdapter.ts:
```diff
@@ -721,6 +721,15 @@ export class TableAdapter extends ModelAdapter {
+    objects.replacePrototypeFunction(Table, '_createTableOrganizer', function(this: Table & { _createTableOrganizerOrig }) {
+      if (this.modelAdapter) {
+        // nop in classic mode
+        return;
```
maybe return null here instead of undefined? But actually makes no difference

bschwarzent replied:
The TableAdapter now sets the organizer to a RemoteTableOrganizer in the _initProperties method.

mvilliger commented on eclipse-scout-core/src/table/ShowInvisibleColumnsFormModel.ts:
```diff
@@ -0,0 +1,97 @@
+              columns: [
+                {
+                  id: 'KeyColumn',
+                  objectType: Column,
```
Specify the generic here (Column)? WidgetMap can handle generics now.

bschwarzent replied:
Updated.

mvilliger commented on eclipse-scout-core/src/table/ShowInvisibleColumnsForm.ts:
```diff
@@ -0,0 +1,80 @@
+    this.insertAfterColumn = null;
+  }
+
+  protected override _jsonModel(): WidgetModel {
```
return type could be a FormModel

bschwarzent replied:
Done.

mvilliger commented on eclipse-scout-core/src/table/ShowInvisibleColumnsForm.ts:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    // Make the columns visible (at their new location)
+    columnsToInsert.forEach(column => column.setVisible(true));
```
each setVisible triggers a 'columnStructureChanged' event. Maybe only one event should be fired after all columns have changed? Or is this intended, that this is executed multiple times?
At least Table#onColumnVisibilityChanged is executed several times which also executes _updateRowWidth() and redraw several times then? Maybe better to only execute this once after all columns have been updated?

bschwarzent replied:
Code was changed to pass "false" as a second argument to the setVisible() method, and manually calls onColumnVisibilityChanged() once afterwards.

mvilliger commented on eclipse-scout-core/src/table/ShowInvisibleColumnsForm.ts:
```diff
@@ -0,0 +1,80 @@
+    if (this.table.tableOrganizer) {
+      return this.table.tableOrganizer.getAddableColumns(this.insertAfterColumn);
+    }
+    return this.table.displayableColumns()
```
Can this case actually happen? This Form is only called from the Organizer, right? Then it must be present? Or not?

bschwarzent replied:
The table organizer is null in classic mode (method _createTableOrganizer is replaced in TableAdapter)

bschwarzent replied:
No longer relevant, logic was moved to TableOrganizer.

mvilliger commented on eclipse-scout-core/src/table/ShowInvisibleColumnsForm.ts:
```diff
@@ -0,0 +1,80 @@
+    this.widget('ColumnsTable').insertRows(rows);
+  }
+
+  protected _getAddableColumns(): Column[] {
```
The list of available columns is computed by the tableOrganizer. Maybe instead of get them here, the organizer could pass it in the model for this form when creating it?

bschwarzent replied:
Agreed. The form now receives the available columns and returns the selected columns via data property. The logic was moved to the table organizer.

mvilliger commented on eclipse-scout-core/src/table/ShowInvisibleColumnsForm.ts:
```diff
@@ -0,0 +1,80 @@
+      return {
+        cells: [
+          scout.create(Cell, {value: column}),
+          this._getColumnTitle(column)
```
In Scout Classic if the tooltip is used as column title, it is displayed using italic font. Do we want this here as well? Or not?

bschwarzent replied:
Not sure how useful this is, but I agree that it's better to have to two implementations to be as similar as possible. Italic font was added.

mvilliger commented on eclipse-scout-core/src/table/TableOrganizer.ts:
```diff
@@ -0,0 +1,175 @@
+   * Returns true if the given column can be modified.
+   */
+  isColumnModifiable(column: Column): boolean {
+    if (this.table.isCustomizable()) {
```
directly return this.table.isCustomizable()

bschwarzent replied:
I left this as it is, so all isColumnXy() methods are structured similarly.

mvilliger commented on eclipse-scout-core/src/table/TableOrganizer.ts:
```diff
@@ -0,0 +1,175 @@
+  protected _removeColumn(column: Column) {
+    column.setVisible(false);
+
+    // TODO bsh [js-table] Should we keep the sorting? In Java, it stays intact (only visible in the organize menu).
```
maybe keep it the same as in java? Or is this expected by the user if a column is removed that the rows change in order?

bschwarzent replied:
Agreed. Code was adjusted.

mvilliger commented:
Table grouping seems not to work correctly. E.g. when grouping by column A and removing column B, the whole grouping seems to be removed. This seems not to be correct?
Furthermore if grouping for multiple columns is configured and one of these is removed, the whole grouping is removed as well?

mvilliger commented:
Another issue:

1. Add an aggregation for a number column (e.g. avg).
2. Remove this number column -> the aggregation row is removed correctly
3. Add the aggregation column again -> aggregation row is not shown, but if the column header is clicked it still shows as active

bschwarzent committed:
Add TableOrganizer and ShowInvisibleColumnsForm for Scout JS

Scout JS tables now support a TableOrganizer, similar to their Java
counterpart. The organizer allows adding, removing and modifying
columns when certain conditions are met. By default, adding and
removing columns changes the "visible" flag of existing columns. In
the future, it will be possible to create new columns via a separate
TableCustomizer (not yet implemented).

The TableHeaderMenu consults the Table about the available organize
actions, which in turn delegates that decision to the TableOrganizer.
The static flags Table#columnAddable, Column#removable and
Column#modifiable can be used to partially disable the organize
operations. To disable table organization completely, set the table
organizer to null. For Scout Classic tables, a special organizer
(RemoteTableOrganizer) is used which lets the table be organized by
the model on the UI server.

When the order or visibility of columns is changed, the artificial
aggregate rows have to be updated accordingly. This is done by
reapplying the grouping in onColumnVisibilityChanged(). The aggregation
values are held in an array that corresponds to the visible columns.
This is now documented at the corresponding arrays, and a jasmine test
was added.

369108

bschwarzent commented:
Thanks for the review. The aggregate rows are now updated correctly.

After additional feedback from @cguglielmo, the `tableOrganizer` property on the table was renamed to `organizer`. The static flags "columnAddable", "removable" and "modifiable" are now also considered in Scout JS. Their default value was change from false to true, but in Scout Classic, they are still defined by the model on the UI server. A special implementation `RemoteTableOrganizer` is used in classic mode.
